### PR TITLE
Rework JPype documentation

### LINKED ISSUE #432: User Guide v2
My coworker feels that our current guide is poorly organized.  It is in random order, doesn't define terms, is *not* compelling, and buries the most important information.

He recommends reordering it something like

- Introduction
  - What and why JPype
  - Use cases examples
  - Limitations
- Basic usage
  - Definition of terms (Class, ClassFactory, wrapper, etc)
  - Importing and Class factories
  - Class, fields, methods, overloading , casting
  - Type conversions
  - Arrays, inner classes
- Advanced usage
  - Threading
  - Proxies
  - Customizers
  - Performance tips

Obviously not something to bite off before 0.7 release, but certainly a goal.

### LINKED ISSUE #433: `__hash__` on Java Objects
Currently ``__hash__`` is defined on a generic Java object type.  As a result any  Java object can be used as a dictionary key even if the object is not immutable.  Not sure how desirable this behavior is.  It appears to be consistent with Java though, as it appears there is no issue storing a list as a Java hash key.   Either way it should be noted in the documentation.

### LINKED ISSUE #455: Document use of autopep8 on JPype
Autopep8 seems to like to reorganize the imports which breaks jpype python files by moving the java imports above the startJVM statement.  This can be avoided by either passing in ``--ignore E402`` or setting the ignore in ``.pep8``.   This should be documented in our userguide someplace.

Example:
```
import jpype
import jpype.imports

jpype.startJVM()

from gov.llnl.math import DoubleArray
```

Result without ignore E402
```
from gov.llnl.math import DoubleArray
import jpype
import jpype.imports

jpype.startJVM()
```

### LINKED ISSUE #463: Debugging tricks
Two tricks that I have been using for joint python/java development are attaching debugging data to Java exceptions and use of serialization to produce core dumps.  As these are very strong tools for working with our code they should appear in our code base somewhere.  However, they are completely unlike other tools in our module so I am not sure how to frame them.  Rather than being so part of the python  module, they are sections of pure Java code that can be used as part of any Java project.  Here are the relevant snippets from my projects internal utility library.

The following code is hereby given the public domain.  If it breaks, it you get to keep all the pieces.

The first is a utility intended to cause a failure in an algorithm to dump some context that lead to the fault into a corefile.  This file can be unloaded in JPype and then examined to see what went wrong.  This is useful when doing large batch runs of a process or capturing a failure at a remote site where the user can send the fault file back to the developer.
```
package example;

import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.io.Serializable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.HashMap;
import java.util.zip.GZIPOutputStream;

/**
 * Utility to simplify the pattern for dumping a core file.
 *
 * To enable core dumps, specify the system property with ".dumpCore" appended
 * java.lang.System.setProperty('org.myproject.dumpCore', 'true');
 *
 * <p>
 * Example:
 * <pre>
 * {@code
 *   package bar;
 *
 *   // Make sure the class and all members are serialiable
 *   class Foo implements Serializable
 *   {
 *      // Be sure to set the version UID so that the methods can be
 *      // altered without breaking the core file
 *      public static final long serialVersionUID=-123727878335L;
 *
 *      public void evaluate(Input input)
 *      {
 *        // Place a try block around the entire body of the function
 *        try {
 *          // something may go wrong here
 *          ...
 *        }
 *
 *        // Catch all exceptions that warrent a core or simply Exception
 *        catch (Exception ex)
 *        {
 *          // Create a core dump file condition on a System property.
 *          CoreDump dump = new CoreDump("org.myproject");
 *
 *          // Add each of the required items to produce the problem
 *          dump.add("Foo", this);
 *          dump.add("input", this.input);
 *          dump.add("exception", ex);
 *
 *          // Each of the items will be saved in the core file
 *          dump.write("foo");
 *
 *          // Rethrow the exception to keep the control flow
 *          throw ex;
 *         }
 *       }
 *    }
 * }
 * </pre>
 *
 */
public class CoreDump
{
  Path coreDirectory = null;
  HashMap<String, Object> core = null;

  /**
   * Create a core dump controlled by a system property.
   *
   * @param property is the system property that should be set to true to enable
   * core dumps.
   */
  public CoreDump(String property)
  {
    try
    {
      String dumpFlag = System.getProperty(property + ".dumpCore");
      if (dumpFlag != null && Boolean.parseBoolean(dumpFlag))
      {
        core = new HashMap<>();
        String coreDir = System.getProperty(property + ".coreDir");
        if (coreDir == null)
        {
          coreDir = ".";
        }
        coreDirectory = Paths.get(coreDir);
      }
    }
    catch (Throwable ex)
    {
    }
  }

  /**
   * Adds a named object to the core dump. This must be called before the call
   * to {@code write}. If items to added to core after it is written, a warning
   * message will be produced.
   *
   * @param key is the name of object to appear in the core.
   * @param obj is the contents of the object.
   * @return the coredump object for argument chaining.
   */
  public CoreDump add(String key, Object obj)
  {
    if (core == null)
      return this;
    try
    {
      core.put(key, obj);
    }
    catch (Throwable ex)
    {
    }
    return this;
  }

  /**
   * Write the core file if core dumps are enabled. This must be the last call
   * to the core dump object.
   *
   * @param key
   * @return true if a core was written.
   *
   */
  public boolean write(String key)
  {
    if (core == null)
      return false;

    try
    {
      int id = (int) Instant.now().getEpochSecond();
      Path file = coreDirectory.resolve(String.format("%s-%08x.ser.gz", key, id));
      if (!Files.exists(this.coreDirectory))
        Files.createDirectories(this.coreDirectory);
      save(file, core);
    }
    catch (Throwable ex)
    {
    }
    core = null;
    return true;
  }

  public void save(Path file, Serializable object) throws IOException
  {
    try (OutputStream fileOut = Files.newOutputStream(file);
            GZIPOutputStream gos = new GZIPOutputStream(fileOut);
            ObjectOutputStream out = new ObjectOutputStream(gos);)
    {
      out.writeObject(object);
    }
  }

}
```

The second is more useful when running an interactively.  When a fault occurs we can get more information from the exception.  Thus when we catch a Java exception in JPype we can get all the variables we need to help debug the problem.

```
package example;

import java.io.Serializable;
import java.util.Map;

/**
 *
 * Utility to add the context that lead to an exception to be thrown.
 * 
 * Example:
 * <pre>
 * {@code
 *     if (Double.isNaN(out[i]))
 *     {
 *       throw Exceptions.builder(new RuntimeException("NaN in output"))
 *               .put("x", x).put("y", y).put("q", q)
 *               .put("inner", inner)
 *               .put("n", nearest(x, q)).get();
 *     }
 * }
 * </pre>
 */
public class Exceptions
{

  /**
   * Get a map containing any attached contexts that have been attached to a
   * throwable.
   *
   * @param throwable is the throwable to probe.
   * @return is a map of strings to objects, it may be empty.
   */
  public static Map<String, Object> getContext(Throwable throwable)
  {
    Throwable t1 = throwable;
    while (t1.getCause() != null)
    {
      t1 = t1.getCause();
      if (t1 instanceof ExceptionContext)
      {
        return ((ExceptionContext) t1).getContext();
      }
    }
    return new ArrayMap<>();
  }

  /**
   * Attach a context to a throwable.
   *
   * @param <Type> is the base type of the throwable.
   * @param throwable is a throwable to attach context to.
   * @param key is a unique key for this object.
   * @param value is the value to be stored.
   * @return the original throwable for chaining or rethrowing.
   */
  public static <Type extends Throwable> Type add(Type throwable, String key, Serializable value)
  {
    Throwable t1 = throwable;
    while (t1.getCause() != null && !(t1 instanceof ExceptionContext))
      t1 = t1.getCause();

    ExceptionContext ex = null;
    if (t1 instanceof ExceptionContext)
    {
      ex = (ExceptionContext) t1;
    }
    else
    {
      ex = new ExceptionContext();
      t1.initCause(ex);
    }

    ex.put(key, value);
    return throwable;
  }

  /**
   * Builder to add more than one item to a context.
   *
   * @param <Type>
   * @param t
   * @return
   */
  public static <Type extends Throwable> ExceptionContextBuilder<Type> builder(Type t)
  {
    return new ExceptionContextBuilder<>(t);
  }

//<editor-fold desc="internal">
  public static class ExceptionContextBuilder<Type extends Throwable>
  {
    Type t;

    ExceptionContextBuilder(Type t)
    {
      this.t = t;
    }

    @SuppressWarnings("ThrowableResultIgnored")
    public ExceptionContextBuilder<Type> put(String key, Serializable value)
    {
      add(t, key, value);
      return this;
    }

    public Type get()
    {
      return t;
    }
  }

  /**
   * Private dummy exception used to store that any context for an exception.
   */
  final static class ExceptionContext extends Throwable
  {
    ArrayMap<String, Object> context = null;

    ExceptionContext()
    {
      // Force the stack trace to be empty because this is not a real exception.
      setStackTrace(new StackTraceElement[0]);
    }

    /**
     * The message to be displayed to the user. Indicates what context 
     * we can probe.
     *
     * @return
     */
    @Override
    public String toString()
    {
      StringBuilder sb = new StringBuilder();
      sb.append("Exception context with contents: ");
      sb.append(StringUtilities.join(context.keySet(), ","));
      return sb.toString();
    }

    /**
     * Add a key value to the context store.
     *
     * @param key
     * @param o
     */
    void put(String key, Serializable o)
    {
      if (context == null)
        context = new ArrayMap<>();
      context.put(key, o);
    }

    /**
     * Get the key value store.
     *
     * @return
     */
    public Map<String, Object> getContext()
    {
      if (context == null)
        context = new ArrayMap<>();
      return context;
    }
  }
//</editor-fold>
}
```

@marscher  I am not quite sure how or where we should place these types of snippets in the current JPype distribution.  I personally find them exceptionally helpful in dealing with problems that crop up while using JPype both locally and remotely.  However, they can't really be compiled into org.jpype codebase as they are pure Java and tying Java code to JPype would be bad.  I am fine with people incorporating and modifying them as needed for their usage.  Any ideas where to place them would be helpful.

### LINKED ISSUE #518: getClassPath after jvm started
Currently we have a few ways of specifying the classpath.  The older direct method was directly specifying 'java.class.path' to the jvm.  It was augmented with a system based off of Matlab in which the classpath is set up with a series of commands prior to calling startJVM.  Things like globbing and pulling from the environment were only in the Matlab based one.  But in JPype 0.7 we modernized by pulling most of the Matlab type features into a keyword args based system.  This new API does not cover all the cases such as construction something using both the environment based path and user options.  So the old Matlab one is still a necessary system and is still quite useful.

But when we added the keyword method, it became much easier to use than the Matlab based one thus it is seeing a lot more use.  However, the kwargs system and the Matlab based system were never designed to work together after the JVM was started.  This leaves the impression that one can use `getClassPath()` to find out the classpath that was used to start the JVM that was specified with the kwargs.  Instead it gives the classpath that would have been used if you did not manually specify it.

The correct procedure has always been to call ``jpype.java.lang.System.getProperty('java.class.path')`` to find out what the loaded classpath was.

It was suggested that rather than always returning the default path (env based with push user specifications) the call to getClassPath should check to see if the JVM is running and if it is get the classpath that was actually used to start the JVM instead of the default one.  It seems reasonable, but I am worried that this will have interactions in the future especially if we have more than one jvm started.  After all if you start a second JVM should it automatically default to the path of the current running one or not?  Alternatively we could just improve the documentation to clarify this issue.

Any opinions on the correct direction for this would be welcomed.

### PR DESCRIPTION
**Pull request by Thrameos**
Okay no getting around this one; eventually we have to update the documentation.  This was the last big task on the JPype 0.8.0 list.  Sorry, this one is going to be a pain to review.

My son is review for clarity and grammar.  Hopefully completed by tomorrow afternoon.

Fixes #432
Fixes #455 
Fixes #433 
Fixes #518 
Fixes #463

### CODE REVIEW

File: doc/userguide.rst [hunk 1]
```diff
@@ -1,259 +1,1509 @@
...
+   :maxdepth: 2
+
+JPype Introduction
+******************
+JPype is a Python module to provide full access to Java from within Python. 
+Unlike Jython, JPype does not achive this by re-implementing Python, but
+instead by interfacing both virtual machines at the native level. This
+shared memory based approach achieves good computing performance, while
+providing the access to the entirety of CPython and Java libraries.
+This approach allows direct memory access between the two machine,
```
Comment (marscher) [inline doc/userguide.rst:15]: machines

File: doc/userguide.rst [hunk 2]
```diff
@@ -1,259 +1,1509 @@
...
+implementation of Java interfaces in Python, and even use of Java threading.
+
+JPype Use Cases
+===============
+
+Here are three typical reasonse to use JPype.
```
Comment (marscher) [inline doc/userguide.rst:21]: reasons

File: doc/userguide.rst [hunk 3]
```diff
@@ -1,259 +1,1509 @@
...
-Eventually, it should be possible to replace Java with Python in many, though
-not all, situations. JSP, Servlets, RMI servers and IDE plugins are all good
-candidates.
+   # Boiler plate stuff to start the module
+   import jpype
+   import jpype.imports
+   from jpype.types import *
 
+   # Launch the JVM
+   jpype.startJVM(classpath=['jars/database.jar'])
```
Comment (marscher) [inline doc/userguide.rst:88]: shouldn't we pass convertStrings here in order to avoid a UserWarning?

Comment (marscher) [reply doc/userguide.rst:88]: Or keep it as simple as possible?

Comment (Thrameos) [reply doc/userguide.rst:88]: The guide was written assuming that we were releasing a 0.8 so some of the deprecated behavior was being removed.

**Comment by Thrameos** [comment]
@marscher okay I believe this is ready for review.

**Pull request merged**